File: Cross-Platform Tests/Java/Amazon/src/test/java/mobileTest.java

import java.io.IOException;
import java.net.MalformedURLException;
import java.util.HashMap;
import java.util.concurrent.TimeUnit;

import org.openqa.selenium.By;
import org.openqa.selenium.remote.DesiredCapabilities;
import org.openqa.selenium.support.ui.ExpectedConditions;
import org.openqa.selenium.support.ui.WebDriverWait;
import org.testng.annotations.AfterTest;
import org.testng.annotations.BeforeTest;
import org.testng.annotations.Parameters;
import org.testng.annotations.Test;
import io.appium.java_client.AppiumDriver;
import io.appium.java_client.TouchAction;
import io.appium.java_client.android.AndroidDriver;
import io.appium.java_client.android.AndroidElement;
import io.appium.java_client.ios.IOSDriver;
import io.appium.java_client.ios.IOSElement;

/**
 * @author Daniel
 * Hybrid test , Mobile test and desktop website.
 * Tested on amazon.co.uk website and amazon app (uk ver).
 */
public class mobileTest {

	@SuppressWarnings("rawtypes")
	private AppiumDriver 	driver;
	private WebDriverWait 	wait; 		//Appium driver wait.
	private String 			os;			//Device operation system / Platform
	private desktopTest 	desktop; 	//desktop test class.
	
	private static final String nativeAPP 	= "NATIVE_APP"; //nativeAPP context 
	private boolean LoggedIn 				= false; 		//user is logged in.
	
	@SuppressWarnings("rawtypes")
	@Parameters({"platformName" , "model" , "deviceName"})
	@BeforeTest
	public void beforeTest(String platformName , String model , String deviceName) throws IOException {
		DesiredCapabilities capabilities = new DesiredCapabilities();
		capabilities.setCapability("user", Properties.user);
		capabilities.setCapability("password", Properties.pass);
		
		//Set this capabilities from the ".XML" file:
		capabilities.setCapability("platformName", platformName);
		capabilities.setCapability("model", model);
		capabilities.setCapability("deviceName", deviceName);
		
		os = platformName;
		if(os.equals("iOS")){
			driver = new IOSDriver(new cloud().getmobileURL(), capabilities);
		}else{
			driver = new AndroidDriver(new cloud().getmobileURL(), capabilities);
		}
		driver.manage().timeouts().implicitlyWait(10 , TimeUnit.SECONDS);
		driver.context(nativeAPP); // Stating test in nativeAPP context.
		wait = new WebDriverWait(driver, 15); 
	}
	
	@Test
  	public void f() {
		boolean mobileSucceed = false;
		try{
			System.out.println("Starting mobile test.");
		  	openAPP();
		  	loginAPP();
		  	search();
		  	mobileSucceed = true;
		}catch(Exception e){ System.out.println(e);}
		
		if(LoggedIn) signOut();
		if(mobileSucceed){
			try{
				System.out.println("Starting desktop test.");
				StartDesktopSession();
				desktop.StartAndLogIn();
				desktop.CartAndCheckout();
				//desktop.CleanBasket(); cleaning basket not necessary.
				desktop.CloseDriver();
			}catch(Exception e){ System.out.println(e);}
		}
  	}

  	@AfterTest
  	public void afterTest() {
  		try{
  			closeAPP();
  			driver.close();
  			driver.quit();
  		}catch(Exception e){
  			System.out.println(e);
  		}
  		
  	}

  	//Open application method.
	private void openAPP(){
		if(driver.getContext().equals(nativeAPP)){
			HashMap<String, Object> script = new HashMap<String, Object>();
			if(os.equals("iOS")){
				script.put("name", Properties.AppNameiOS);
			}
			else{
				script.put("name", Properties.AppNameAndroid);
			}
			driver.executeScript("mobile:application:open", script);
		}
	}
	
	//Login method. 
	private void loginAPP(){
		try{
			if(os.equals("iOS")){
				wait.until(ExpectedConditions.elementToBeClickable(driver.findElement(By.xpath(TestObjects.Sign_in)))).click();
				driver.findElement(By.xpath(TestObjects.userName)).sendKeys(Properties.AppUser);
				driver.findElement(By.xpath(TestObjects.password)).sendKeys(Properties.AppPass);
				driver.findElement(By.xpath(TestObjects.login_button)).click();
			}
			else{//Android require click on the fields.
				wait.until(ExpectedConditions.elementToBeClickable(driver.findElement(By.xpath(TestObjects.Sign_in)))).click();
				driver.findElement(By.xpath(TestObjects.userName)).click();
				driver.findElement(By.xpath(TestObjects.userName)).sendKeys(Properties.AppUser);
				driver.findElement(By.xpath(TestObjects.password)).click();
				driver.findElement(By.xpath(TestObjects.password)).sendKeys(Properties.AppPass);
				driver.findElement(By.xpath(TestObjects.Android_loginBTN)).click();
			}
			
			this.LoggedIn = true;
		}catch(Exception e){
			System.out.println(e);
		}
	}
	
	//Search method.
	private void search(){
		boolean menu_loaded = false;
		if(os.equals("iOS")){
			try{//Finding search area could be problematic therefore trying twice. 
				wait.until(ExpectedConditions.visibilityOfElementLocated(By.xpath(TestObjects.Search))).sendKeys(Properties.SearchValue);
				menu_loaded = true;
			}catch(Exception e){}
			if(!menu_loaded){
				wait.until(ExpectedConditions.visibilityOfElementLocated(By.xpath(TestObjects.Search))).sendKeys(Properties.SearchValue);
			}
		}
		else{
			wait.until(ExpectedConditions.visibilityOfElementLocated(By.xpath(TestObjects.Search))).click();
			wait.until(ExpectedConditions.visibilityOfElementLocated(By.xpath(TestObjects.AndroidSearch2))).sendKeys(Properties.SearchValue);
		}
		
		driver.findElement(By.xpath(TestObjects.CDsAndVinyl)).click();
		driver.findElement(By.xpath(TestObjects.DarkSideCD)).click();
		//This part of code separate two cases of ScrollTO method.
		//In order to find Basket button scrolling down and search for the text "Add to Basket" .
		try{
			if(os.equals("iOS")){
				driver.findElementByClassName("UIAButton");
				//Scrolling down and find element by it's tag name.
				IOSElement tbl = (IOSElement) driver.findElementByClassName("UIAWebView");
				tbl.scrollTo("Add to Basket").click();
			}
			else{
					driver.scrollTo("Add to Basket");
					driver.findElement(By.xpath(TestObjects.Android_AddtoBasket)).click();
			}
		}catch(Exception e){
			System.out.println(e);
		}
	}
	
	//signOut method.
	private void signOut(){
		openAPP();//Reopen the application  in order to back the main menu.
		try{
			driver.findElement(By.xpath(TestObjects.AppMenu)).click();
			if(os.equals("iOS")){
				driver.findElement(By.xpath(TestObjects.SignOut)).click();
				driver.findElement(By.xpath(TestObjects.PopUpSignOut)).click();
			}
			else{
				driver.scrollTo("Not "+Properties.NameOnly+"? Sign out");
				driver.findElement(By.xpath("//*[@text='Not "+Properties.NameOnly+"? Sign out']")).click();
				driver.findElement(By.xpath(TestObjects.Android_PopUpSignOut)).click();
			}
			this.LoggedIn = false;
		}catch(Exception e){
			System.out.println("Sign out unsuccessfully.");
		}
	}
	
	private void closeAPP(){
		if(driver.getContext().equals(nativeAPP)){
			HashMap<String, Object> script = new HashMap<String, Object>();
			if(os.equals("iOS")){
				script.put("name", Properties.AppNameiOS);
			}
			else{
				script.put("name", Properties.AppNameAndroid);
			}
			driver.executeScript("mobile:application:close", script);
		}
	}
	
	//Setting the desktop test to be ready.
	private void StartDesktopSession() throws MalformedURLException{
		desktop = new desktopTest();
		desktop.beforeTest();
	}
}

File: Cross-Platforms tests/Python/Amazon - CrossPlatform Python/src/MobileTest.py
from appium import webdriver

import TestParameters
from AppObjects import Locators
from DesktopTest import desktoptest
from selenium.webdriver.common.by import By
from selenium.webdriver.support.wait import WebDriverWait
from selenium.webdriver.support import expected_conditions as EC

"""
This is an implementation of cross driver test on amazon mobile app and website.
Using Appium on mobile application and Selenium on the website session.
In both cases (App and website) all the Locators been tested on the UK version.
The test move between Amazon Application (Adding to basket an item scenario) and then fill information on Amazon's UK website.
All the test locators can be found at AppObjects.py.
Personal information supplied at TestParameters.py.
"""
class MobileTest():
    
    def __init__(self):
        self.driver = self.create_AppiumDriver()
        self.driver.implicitly_wait(15)
        self.driver.switch_to.context("NATIVE_APP")
        self.wait = WebDriverWait(self.driver , 10)
        
    def create_AppiumDriver(self):
        cloud = 'https://' + TestParameters.host + '/nexperience/perfectomobile/wd/hub'
        return webdriver.Remote(cloud ,TestParameters.iPhone_6)
    
    """
    Test session , here comes all the methods of the test by order.
    """
    def TEST(self):
        try:
            self.openApp()
            self.LoginApp()
            self.search()
            self.LogOutAPP()
            self.closeApp() 
            
        except Exception as e:
            print e
            
        finally:
            self.EndTest()
        
        #From that point switching to Selenium driver and starting web test.
        try:
            desktop_test = desktoptest()
            desktop_test.start_and_login()
            desktop_test.go_cart_checkout()
            desktop_test.clear_basket()
            desktop_test.end_test()
        except Exception as e:
            print e
    
    """       
    Test Methods 
    """
    def openApp(self):
        script = {"name" : "Amazon"}
        self.driver.execute_script('mobile:application:open' , script)
        
    def closeApp(self):
        script = {"name" : "Amazon"}
        self.driver.execute_script('mobile:application:close' , script)
        
    def EndTest(self):
        try:
            self.driver.close()
            self.driver.quit()
        except Exception as e:
            print e
    
    
    def LoginApp(self):
        try:
            self.wait.until(EC.visibility_of_element_located((By.XPATH , Locators.Sign_in))).click()
            self.driver.find_element_by_xpath(Locators.userName).send_keys(TestParameters.AppUser)
            self.driver.find_element_by_xpath(Locators.password).send_keys(TestParameters.AppPass)
            self.driver.find_element_by_xpath(Locators.login_btn_ios).click()
        except Exception as e:
            print e    
    
    def LogOutAPP(self):
        try:
            self.driver.find_element_by_xpath(Locators.AppMenu).click()
            self.driver.find_element_by_xpath("//*[@name = 'Not "+TestParameters.YourName+"? Sign out' or @text= 'Not" + TestParameters.YourName +"? Sign out']").click()
            self.driver.find_element_by_xpath(Locators.Sign_out_popup_ios).click()
        except Exception as e:
            print e
    
    def search(self):
        try: 
            try:
                self.wait.until(EC.visibility_of_element_located((By.XPATH , Locators.search_box))).send_keys(TestParameters.searchValue)
            except:
                self.wait.until(EC.visibility_of_element_located((By.XPATH , Locators.search_box))).send_keys(TestParameters.searchValue)
               
            self.driver.find_element_by_xpath(Locators.search_result).click()
            self.driver.find_element_by_xpath(Locators.darkSideAlbum).click()
            #In order to perform a scroll choose two elements.
            #Then driver.scroll method scrolling from the first element to the second element. 
            elem = self.driver.find_element_by_xpath("//*[@label = 'Details']") 
            elem2 = self.driver.find_element_by_xpath("//*[@label = 'Original recording remastered, Original recording reissued']")
            self.driver.scroll(elem, elem2)
            self.driver.find_element_by_xpath(Locators.AddtoBasket).click()
            
        except Exception as e:
            print e
    

""" Main method """
if __name__ == '__main__':
    MobileTest().TEST()

